std/range/package.d

```diff
@@ -4125,32 +4125,60 @@ pure unittest
     Generate lockstep's opApply function as a mixin string.
     If withIndex is true prepend a size_t index to the delegate.
 */
-private string lockstepMixin(Ranges...)(bool withIndex)
+private string lockstepMixin(Ranges...)(bool withIndex, bool reverse)
 {
     import std.format : format;
 
     string[] params;
     string[] emptyChecks;
     string[] dgArgs;
     string[] popFronts;
+    string indexDef;
+    string indexInc;
 
     if (withIndex)
     {
         params ~= "size_t";
         dgArgs ~= "index";
+        if (reverse)
+        {
+            indexDef = q{
+                size_t index = ranges[0].length-1;
+                enforce(_stoppingPolicy == StoppingPolicy.requireSameLength, "lockstep can only be used with foreach_reverse when stoppingPolicy == requireSameLength");
```

Comment:
> yep but only for bidirectional ranges + foreach_reverse.

Only for bidirectional ranges + foreach_reverse \*with an index\*. That means you get a hidden cost of who-knows-what only in a very rare corner case. That's certainly breaking the principle of least astonishment.

This is the current behavior:

foreach_reverse (e1, e2; lockstep([1,2,3], [4,5])) {
    assert(e1 == e2 - 2);
}

That leaves off the 1 from the first array. If that's not an array, but e.g. some bidirectional generator, I'd still expect it to iterate from the end forward. Since backward iteration without an index doesn't require hasLength, it's not possible to have the behavior you describe without the index. So either the behavior is different when using an index, or we simply disallow the cases where the behavior would differ. I'm very strongly in favor of the latter.